elki-outlier/src/main/java/elki/outlier/density/HySortOD.java

```diff
@@ -0,0 +1,694 @@
+        return;
+
+      // Get the first value from the given range (minRowIdx, maxRowIdx)
+      int value = this.H.get(parent.begin).getCoordAt(col);
+
+      // Initialise the next range
+      int begin = parent.begin;
+      int end = -1;
+
+      // map the values in the current range
+      int i = parent.begin;
+      for(; i <= parent.end; i++) {
+
+        // when the value change the node is created
+        if(this.H.get(i).getCoordAt(col) != value) {
+
+          // mark the end of the current value
+          end = i - 1;
+
+          // create node for 'value' in 'col'
+          Node child = new Node(value, begin, end);
+          parent.add(child);
+
+          // map child values in the next dimension
+          buildIndex(child, col + 1);
+
+          // start new range
+          begin = i;
+
+          // update value
+          value = this.H.get(i).getCoordAt(col);
+        }
+      }
+
+      // map last value
+      end = i - 1;
+      Node child = new Node(value, begin, end);
+      parent.add(child);
+
+      buildIndex(child, col + 1);
+    }
+
+    @Override
+    int[] getDensities() {
+      int n = H.size();
+      int[] W = new int[n];
+
+      for(int i = 0; i < n; i++) {
+        W[i] = density(i, root, 0);
+        Wmax = Math.max(Wmax, W[i]);
+      }
+
+      return W;
+    }
+
+    /**
+     * Recursive computing of density.
+     * 
+     * @param i Index of the hypercube
+     * @param parent Node of the tree
+     * @param col Coordinate of the hypercube
+     * @return Density
+     */
+    private int density(int i, Node parent, int col) {
+      int density = 0;
+
+      if(parent.childs.isEmpty()) {
+        for(int k = parent.begin; k <= parent.end; k++) {
+          if(isImmediate(this.H.get(i), this.H.get(k))) {
+            density += H.get(k).getDensity();
+          }
+        }
+      }
+      else {
+
+        int lftVal = this.H.get(i).getCoordAt(col) - 1;
+        int midVal = this.H.get(i).getCoordAt(col);
+        int rgtVal = this.H.get(i).getCoordAt(col) + 1;
+
+        Node lftNode = parent.childs.get(lftVal);
+        Node midNode = parent.childs.get(midVal);
+        Node rgtNode = parent.childs.get(rgtVal);
+
+        int nextCol = Math.min(col + 1, this.H.get(i).getNumDimensions() - 1);
+
+        if(Objects.nonNull(lftNode)) {
+          density += density(i, lftNode, nextCol);
+        }
+        if(Objects.nonNull(midNode)) {
+          density += density(i, midNode, nextCol);
+        }
+        if(Objects.nonNull(rgtNode)) {
+          density += density(i, rgtNode, nextCol);
+        }
+      }
+
+      return density;
+    }
+
+    /**
+     * Tree node.
+     * 
+     * @author Cabral, Eugênio F.
+     *
+     */
+    private class Node {
+
+      /**
+       * Index information.
+       */
+      public final int value, begin, end;
+
+      /**
+       * Childs of the node.
+       */
+      public final HashMap<Integer, Node> childs;
```

Comment:
Using a Int2ObjectOpenHashMap<Node> avoids boxing overheads.
But it may be worth instantiating the map only when used.

Reply:
Also childs -> children